### PR TITLE
Add --legacy-packages-directory to Xplat restore

### PR DESCRIPTION
**Pull request by joelverhagen**
Add a new switch `--legacy-packages-directory` to `NuGet.CommandLine.Xplat` (i.e. `dotnet restore`) which allows the caller to restore to the packages folder folders with original case ID and version.

Since the correct format of the global packages folder is `{root}/{lowercase-id}/{lowercase-version}`, this is done by first restoring to lowercase folder names then, as a post-process step to the restore operation, copy the packages to `{root}/{original-id}/{original-version}`. This allows the legacy mode to be minimally invasive to the main restore operation.

This command line option is part of a backwards compatibility story so that 1.0.0 Corehost can still find packages restored by newer version of NuGet.

Corehost reads a `.deps.json` file generated by .NET CLI. .NET CLI generates the `.deps.json` file from the `project.lock.json` file that we write out. When the `--legacy-packages-directory` switch is passed, the lock file written out has the `path` property of each `package` library converted from lowercase to original case.

Next steps are modifying CLI to include the `path` property in the `.deps.json` file and modifying Corehost to prefer this `path` property over the concatenation of the library ID + library version.

@emgarten @alpaix @rohit21agrawal @rrelyea

### CODE REVIEW

File: src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/RestoreCommand.cs [hunk 1]
```diff
@@ -88,6 +94,8 @@ internal static class RestoreCommand
 
                     using (var cacheContext = new SourceCacheContext())
                     {
+                        var legacyPackagesDirectory = legacyPackagesDirectoryOption.HasValue();
```
Comment (alpaix) [inline src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/RestoreCommand.cs]: nit. `isLegacyPackagesDirectory`

File: src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/RestoreCommand.cs [hunk 2]
```diff
@@ -100,13 +108,14 @@ internal static class RestoreCommand
...
 
-                        var restoreContext = new RestoreArgs()
+                        var restoreContext = new RestoreArgs
                         {
                             CacheContext = cacheContext,
                             LockFileVersion = LockFileFormat.Version,
                             ConfigFile = configFile.HasValue() ? configFile.Value() : null,
                             DisableParallel = disableParallel.HasValue(),
                             GlobalPackagesFolder = packagesDirectory.HasValue() ? packagesDirectory.Value() : null,
+                            LowercaseGlobalPackagesFolder = !legacyPackagesDirectory,
```
Comment (rohit21agrawal) [inline src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/RestoreCommand.cs]: IsLowerCaseGlobalPackagesFolder

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/OriginalCaseGlobalPackageFolder.cs [hunk 3]
```diff
@@ -0,0 +1,159 @@
+﻿using System;
```
Comment (alpaix) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/OriginalCaseGlobalPackageFolder.cs]: header

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/OriginalCaseGlobalPackageFolder.cs [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+                            Strings.Log_ConvertedPackageToOriginalCase,
+                            identity));
+                    }
+                }
+            }
+        }
+
+        public void ConvertLockFileToOriginalCase(LockFile lockFile)
+        {
+            foreach (var library in lockFile.Libraries)
```
Comment (alpaix) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/OriginalCaseGlobalPackageFolder.cs]: nit: `.Where(library.Type == LibraryType.Package` for brevity

Comment (joelverhagen) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/OriginalCaseGlobalPackageFolder.cs]: Fixed.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/OriginalCaseGlobalPackageFolder.cs (hunk 3 above)
Comment (joelverhagen) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/OriginalCaseGlobalPackageFolder.cs]: Fixed.

File: src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/RestoreCommand.cs (hunk 2 above)
Comment (joelverhagen) [inline src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/RestoreCommand.cs]: Fixed.

File: src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/RestoreCommand.cs (hunk 1 above)
Comment (joelverhagen) [inline src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/RestoreCommand.cs]: Fixed.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/OriginalCaseGlobalPackageFolder.cs [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+            var localPackage = NuGetv3LocalRepositoryUtility.GetPackage(
+                _localRepositories,
+                library.Name,
+                library.Version);
+
+            if (localPackage != null && File.Exists(localPackage.Package.ZipPath))
+            {
+                using (var stream = File.OpenRead(localPackage.Package.ZipPath))
+                {
+                    await stream.CopyToAsync(destination, bufferSize: 4096, cancellationToken: token);
```
Comment (alpaix) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/OriginalCaseGlobalPackageFolder.cs:157]: From what I've learned it won't copy async unless you specify `useAsync` in stream constructor

Comment (joelverhagen) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/OriginalCaseGlobalPackageFolder.cs:157]: Yeah, I think you're right. We have other places that use `File.OpenRead` so I doubt it's an issue. Will fix anyhow.

**Comment by emgarten** [comment]
xplat restore is dead :skull:

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/OriginalCaseGlobalPackageFolder.cs (hunk 5 above)
Comment (joelverhagen) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/OriginalCaseGlobalPackageFolder.cs:157]: Fixed.

**Comment by alpaix** [comment]
lgtm :shipit:

**Comment by emgarten** [comment]
Looks great! :clap: :shipit:

**Pull request merged**